Add an INACTIVE CheckpointPhase

tjwatson opened the pull request:
Add an INACTIVE CheckpointPhase
    
Change CheckpointPhase.getPhase to never return null.
If no checkpoint is being done then getPhase will return
the INACTIVE phase.  Always register the CheckpointPhase
service.  This service is internal so is not beta gaurded from
registration.
    
Updates to all places that checked for null instead of
the INACTIVE phase.

tjwatson committed:
Add an INACTIVE CheckpointPhase

Change CheckpointPhase.getPhase to never return null.
If no checkpoint is being done then getPhase will return
the INACTIVE phase.  Always register the CheckpointPhase
service.  This service is internal so is not beta gaurded from
registration.

Updates to all places that checked for null instead of
the INACTIVE phase.

tjwatson commented:
#build

tjwatson committed:
Add a static method to get the simple name from any TestName

Sometimes other tests that repeat need to get the simple method name.
This static allows them to use the same logic to strip out the repeating name.

tjwatson committed:
Fix issues with finding the test method name.

The componenttest.topology.utils.FATServletClient.getTestMethodSimpleName(TestName)
sometimes returns the class-name.method-name.

This fixes that plus fixes an issue with the LocalEJBTest
that was having issues finding the proper TestMethod.

io.openliberty.checkpoint.fat.FATSuite.getTestMethod(Class<T>, TestName)
was updated to fail the test if a proper TestMethod enum
could not be found.

tjwatson commented:
#build

Azquelt commented on dev/com.ibm.ws.kernel.boot.core/src/io/openliberty/checkpoint/spi/CheckpointPhase.java:
```diff
@@ -39,13 +39,22 @@ public enum CheckpointPhase {
      */
-    DEPLOYMENT;
+    DEPLOYMENT,
+    /**
+     * Phase that indicates there is no checkpoint being done.
+     */
+    INACTIVE(true);
```
Update Javadoc for `CheckpointPhase` which currently says
> This enum is registered as a service _when a checkpoint operation is being done_.

tjwatson replied:
I've rebased to latest integration branch so you comment is not included in the change view.  But I updated the class doc to:

```
Phase which a checkpoint of the running process is being taken. 
This enum is registered as a service when running in an OSGi framework.
```

Azquelt commented on dev/com.ibm.ws.kernel.boot.core/src/io/openliberty/checkpoint/spi/CheckpointPhase.java:
```diff
@@ -124,7 +149,7 @@ final public boolean restored() {
      *
      * @param hook the hook to be used to prepare and restore
      *                 a checkpoint process.
-     * @throws IllegalStateException if the phase is in progress.
+     * @throws IllegalStateException if the phase is not in progress.
      * @see CheckpointHook#MULTI_THREADED_HOOK
```
This javadoc also needs to be updated to say what the return value means (same for `addMultiThreadedHook()`)

tjwatson replied:
Thanks for pointing that out.  Here is the update I made which is applied to both methods:

```java
    /**
     * Adds a {@code CheckpointHook} for this checkpoint phase.
     * The hook will be run to {@link CheckpointHook#prepare() prepare}
     * and {@link CheckpointHook#restore() restore} a checkpoint
     * process while the JVM is in multi-threaded mode. If the
     * checkpoint process is already {@link #restored() restored}
     * then the hook is not added and {@code false} is returned,
     * indicating that the hook will not be called.
     * A return value of {@code true} indicates that the hook will
     * be called for both {@link CheckpointHook#prepare() prepare}
     * and {@link CheckpointHook#restore() restore}.
     *
     * @param hook the hook to be used to prepare and restore
     *                 a checkpoint process.
     * @throws IllegalStateException if the phase is not in progress.
     * @return true if the hook is successfully added; otherwise false
     *         is returned.
     * @see CheckpointHook#MULTI_THREADED_HOOK
     */
```

tjwatson committed:
Remove null checks for the checkpoint phase.

With changes to add the INACTIVE checkpoint phase
there is no need to check for null from the
CheckpointPhase.getPhase() method.

tjwatson commented:
#build

tjwatson commented:
@tkburroughs I included you to review because you were an early adopter of the original service design.  I changed the EJB code to no longer consume the phase as a service because it simplifies the code a bit.

tkburroughs commented on dev/com.ibm.ws.ejbcontainer/src/com/ibm/ws/ejbcontainer/osgi/internal/EJBRuntimeImpl.java:
```diff
@@ -281,7 +286,9 @@ public EJBRuntimeImpl(@Reference(cardinality = ReferenceCardinality.OPTIONAL) Ch
     protected final void checkpointRestored(ServiceReference<?> checkpoint) {
         // Resume all non-persistent timers on checkpoint restore
-        TimerNpRunnable.resume();
+        if (checkpointPhase != CheckpointPhase.INACTIVE) {
```
While the code is not wrong, is it unnecessary? Curious why checkpointRestored() would ever be called if INACTIVE?

tjwatson replied:
We now always register the `CheckpointPhase` that is returned by the static method `CheckpointPhase.getPhase()`.  For the case where there is no `checkpoint` action being done this phase is going to be the `INACTIVE` phase which reports to already be `restored`.  So the target filter will match and this method will get called during activation.

We could instead update the target filter to be:

```
target = "(&(" + CheckpointPhase.CHECKPOINT_RESTORED_PROPERTY + "=true)" +
         "(!(" + CheckpointPhase.CHECKPOINT_PROPERTY + "=INACTIVE)))"
```
That should allow us to call `TimerNpRunnable.resume()` unconditionally here.

tkburroughs replied:
Ah, got it; the match is on the `restored` setting in the phase.  Now that I understand, I'm fine with the code as is; or with the target filter change.

tkburroughs commented on dev/com.ibm.ws.ejbcontainer/src/com/ibm/ws/ejbcontainer/osgi/internal/EJBRuntimeImpl.java:
```diff
@@ -260,16 +260,21 @@ private static RuntimeException rethrow(Throwable t) {
 
         // For any Checkpoint phase, pause all non-persistent timers until checkpoint restored
-        if (checkpointPhase != null && !checkpointPhase.restored()) {
+        if (!checkpointPhase.restored()) {
```
This is a change in behavior, as previously the code was using the `checkpointPhase` parameter and not `this.checkpointPhase`. Therefore, the code would have ignored the test property (now testPhase).  The issue is that when just using the test property, there is no mechanism to call `checkpointRestored`, so timers will never run.

tjwatson replied:
The check for `phase != CheckpointPhase.INACTIVE` can be thought of as a replacement for `phase != null`.  

So I think the same code could have been:

```java
        CheckpointPhase checkpointPhase = CheckpointPhase.getPhase();
        if (checkpointPhase != CheckpointPhase.INACTIVE || !ProductInfo.getBetaEdition()) {
            this.checkpointPhase = checkpointPhase;
        } else {
            CheckpointPhase testPhase = CheckpointPhase.getPhase(System.getProperty("io.openliberty.ejb.checkpoint.phase", ""));
            this.checkpointPhase = testPhase != null ? testPhase : checkpointPhase;
        }

        // For any Checkpoint phase, pause all non-persistent timers until checkpoint restored
        if (checkpointPhase != CheckpointPhase.INACTIVE && !checkpointPhase.restored()) {
            TimerNpRunnable.pause();
        }
    }
```
I can change it to that, but I think it is equivalent to the code I have in the PR

tkburroughs replied:
The old behavior would not call `pause()` if checkpoint phase was obtained from the property, rather than the real checkpoint service.  Because `checkpoint` is no longer passed in as a parameter, `checkpoint` is the same as `this.checkpoint`, whereas in the original code it was not.  In the original code, `checkpoint` could have been `null`, whereas `this.checkpoint` may have been APPLICATION (set from the property).

The above code is still not the same behavior as before.... instead, it would need to be something like

```
        boolean checkpointFromProperty = false;
        CheckpointPhase checkpointPhase = CheckpointPhase.getPhase();
        if (checkpointPhase != CheckpointPhase.INACTIVE || !ProductInfo.getBetaEdition()) {
            this.checkpointPhase = checkpointPhase;
        } else {
            checkpointFromProperty = true;
            this.checkpointPhase = CheckpointPhase.getPhase(System.getProperty("io.openliberty.ejb.checkpoint.phase", ""));
        }

        // For any Checkpoint phase, pause all non-persistent timers until checkpoint restored
        if (!checkpointFromProperty && !checkpointPhase.restored()) {
            TimerNpRunnable.pause();
        }
    }
```

tjwatson replied:
Ah, now I get it.  But with the change to how `INACTIVE` is always registered then we always do call `checkpointRestored` and this constructor would have set `this.checkpointPhase` to the phase specified by the property which the tests would have set it to FEATURES, DEPLOYMENT or APPLICATIONS so the check for INACTIVE would pass making it call the `TimerNpRunnable.resume()`

This is a bit obscure.  I'll try to make this more straightforward.

tjwatson commented:
#libby
